merkletree/merkletree.go

```diff
@@ -192,6 +184,30 @@ insertLoop:
 	}
 }
 
+// visits all leaf-nodes and calls callBack on each of them
+// doesn't modify the underlying tree m
+func (m *MerkleTree) visitLeafNodes(callBack func(*userLeafNode)) {
+	visitULNsInternal(m.root, callBack)
+}
+
+func visitULNsInternal(nodePtr interface{}, callBack func(*userLeafNode)) {
```

Comment:
I would change `nodePtr interface{}` into `nodePtr MerkleNode`. I also want to change L43 and L117, too. So if it makes sense to you ...